usermods/sht/usermod_sht.h

```diff
@@ -0,0 +1,312 @@
+      *
+      * Tips:
+      * 1. You can use "if (WLED_CONNECTED)" to check for a successful network connection.
+      *    Additionally, "if (WLED_MQTT_CONNECTED)" is available to check for a connection to an MQTT broker.
+      *
+      * 2. Try to avoid using the delay() function. NEVER use delays longer than 10 milliseconds.
+      *    Instead, use a timer check as shown here.
+      */
+    void loop()
+    {
+      if (!enabled || !initDone || strip.isUpdating()) return;
+
+      if (isShtReady()) {
+        if (millis() - shtLastTimeUpdated > 30000 && !shtDataRequested) {
+          shtTempHumidSensor->requestData();
+          shtDataRequested = true;
+
+          shtLastTimeUpdated = millis();
+        }
+
+        if (shtDataRequested) {
+          if (shtTempHumidSensor->dataReady()) {
+            if (shtTempHumidSensor->readData(false)) {
+              shtCurrentTemp = shtTempHumidSensor->getTemperature();
+              shtCurrentHumidity = shtTempHumidSensor->getHumidity();
+
+              publishTemperatureAndHumidityViaMqtt();
+              shtReadDataSuccess = true;
+            }
+            else {
+              shtReadDataSuccess = false;
+            }
+
+            shtDataRequested = false;
+          }
+        }
+      }
+    }
+
+    void onMqttConnect(bool sessionPresent) {
+      if (haMqttDiscovery) publishHomeAssistantAutodiscovery();
+    }
+
+    void addToConfig(JsonObject &root)
+    {
+      JsonObject top = root.createNestedObject(FPSTR(_name)); // usermodname
+
+      top[FPSTR(_enabled)] = enabled;
+      top[FPSTR(_haMqttDiscovery)] = haMqttDiscovery;
+    }
+
+    /**
+     * readFromConfig() is called before setup() to populate properties from values stored in cfg.json
+     *
+     * The function should return true if configuration was successfully loaded or false if there was no configuration.
+     */
+    bool readFromConfig(JsonObject &root)
+    {
+      JsonObject top = root[FPSTR(_name)];
+      if (top.isNull()) {
+        DEBUG_PRINTF("[%s] No config found. (Using defaults.)\n", _name);
+        return false;
+      }
+
+      bool oldEnabled = enabled;
+      bool oldHaMqttDiscovery = haMqttDiscovery;
+      
+      getJsonValue(top[FPSTR(_enabled)], enabled);
+      getJsonValue(top[FPSTR(_haMqttDiscovery)], haMqttDiscovery);
+
+      // First run: reading from cfg.json, nothing to do here, will be all done in setup()
+      if (!firstRunDone) {
+        DEBUG_PRINTF("[%s] First run, nothing to do\n", _name);
+      }
+      // Check if mod has been en-/disabled
+      else if (enabled != oldEnabled) {
+        enabled ? setup() : cleanup();
+        DEBUG_PRINTF("[%s] Usermod has been en-/disabled\n", _name);
+      }
+      // Config has been changed, so adopt to changes
+      else if (enabled) {
+        if (oldHaMqttDiscovery != haMqttDiscovery && haMqttDiscovery) {
+          publishHomeAssistantAutodiscovery();
```

Comment:
At this stage (during boot invocation) WiFi may not yet be connected.

Reply:
Hrrmm, buuuut: `publishHomeAssistantAutodiscovery` checks for `WLED_MQTT_CONNECTED`, which implies network availability, right? :)

I mostly tend to have methods checking their requirements themselves, so I can just call them and they check themselves if they can safely execute.